test/e2e/antreaipam_test.go

```diff
@@ -156,29 +200,215 @@ func testAntreaIPAMPodConnectivityDifferentNodes(t *testing.T, data *TestData) {
+		t.Fatalf("Error when waiting for StatefulSet Pods to get IPs: %v", err)
+	}
+	checkStatefulSetIPPoolAllocation(t, data, stsName, testAntreaIPAMNamespace, ipPoolName, ipOffsets, reservedIPOffsets)
+
+	cleanup()
+	checkStatefulSetIPPoolAllocation(t, data, stsName, testAntreaIPAMNamespace, ipPoolName, nil, nil)
+}
+
+func checkStatefulSetIPPoolAllocation(tb testing.TB, data *TestData, name string, namespace string, ipPoolName string, ipOffsets, reservedIPOffsets []int32) {
+	ipPool, err := data.crdClient.CrdV1alpha2().IPPools().Get(context.TODO(), ipPoolName, metav1.GetOptions{})
+	if err != nil {
+		tb.Fatalf("Failed to get IPPool %s, err: %+v", ipPoolName, err)
+	}
+	startIP := net.ParseIP(ipPool.Spec.IPRanges[0].Start)
+	expectedIPAddressMap := map[string]*crdv1alpha2.IPAddressState{}
+	for i, offset := range ipOffsets {
+		ipString := utilnet.AddIPOffset(utilnet.BigForIP(startIP), int(offset)).String()
+		podName := fmt.Sprintf("%s-%d", name, i)
+		expectedIPAddressMap[ipString] = &crdv1alpha2.IPAddressState{
+			IPAddress: ipString,
+			Phase:     crdv1alpha2.IPAddressPhaseAllocated,
+			Owner: crdv1alpha2.IPAddressOwner{
+				Pod: &crdv1alpha2.PodOwner{
+					Name:        podName,
+					Namespace:   namespace,
+					ContainerID: "",
+				},
+			},
+		}
+	}
+	for i, offset := range reservedIPOffsets {
+		ipString := utilnet.AddIPOffset(utilnet.BigForIP(startIP), int(offset)).String()
+		stsOwner := &crdv1alpha2.StatefulSetOwner{
+			Name:      name,
+			Namespace: namespace,
+			Index:     i,
+		}
+		if _, ok := expectedIPAddressMap[ipString]; ok {
+			expectedIPAddressMap[ipString].Owner.StatefulSet = stsOwner
+		} else {
+			expectedIPAddressMap[ipString] = &crdv1alpha2.IPAddressState{
+				IPAddress: ipString,
+				Phase:     crdv1alpha2.IPAddressPhaseReserved,
+				Owner: crdv1alpha2.IPAddressOwner{
+					StatefulSet: stsOwner,
+				},
+			}
+		}
+	}
+	expectedIPAddressJson, _ := json.Marshal(expectedIPAddressMap)
+	tb.Logf("expectedIPAddressMap: %s", expectedIPAddressJson)
+
+	err = wait.Poll(time.Second*5, defaultTimeout, func() (bool, error) {
+		ipPool, err := data.crdClient.CrdV1alpha2().IPPools().Get(context.TODO(), ipPoolName, metav1.GetOptions{})
+		if err != nil {
+			tb.Fatalf("Failed to get IPPool %s, err: %+v", ipPoolName, err)
+		}
+		actualIPAddressMap := map[string]*crdv1alpha2.IPAddressState{}
+	actualIPAddressLoop:
+		for i, ipAddress := range ipPool.Status.IPAddresses {
+			for expectedIP := range expectedIPAddressMap {
+				if ipAddress.IPAddress == expectedIP {
+					actualIPAddressMap[expectedIP] = ipAddress.DeepCopy()
+					if actualIPAddressMap[expectedIP].Owner.Pod != nil {
+						actualIPAddressMap[expectedIP].Owner.Pod.ContainerID = ""
+					}
+					continue actualIPAddressLoop
+				}
+			}
+			if ipAddress.Owner.Pod != nil && ipAddress.Owner.Pod.Namespace == namespace && strings.HasPrefix(ipAddress.Owner.Pod.Name, name) {
+				actualIPAddressMap[ipAddress.IPAddress] = &ipPool.Status.IPAddresses[i]
```

Comment:
isn't `ipPool.Status.IPAddresses[i]` the same as `ipAddress`?

Reply:
No, `ipAddress` is a temporary object in this loop, thus we shouldn't use its pointer outside the loop.

Reply:
Got it, thanks